drivers/misc/stm32_tsc/stm32_tsc.c

```diff
@@ -0,0 +1,332 @@
+			LOG_ERR("%s: group %d is used as shield, but has invalid channel I/Os. "
+				"Can only have one",
+				dev->name, group->group);
+			return -EINVAL;
+		}
+
+		/* clear schmitt trigger hysteresis for enabled I/Os */
+		sys_clear_bits((mem_addr_t)&config->tsc->IOHCR,
+			       GET_GROUP_BITS(group->channel_ios | group->sampling_io));
+
+		/* set channel I/Os */
+		sys_set_bits((mem_addr_t)&config->tsc->IOCCR, GET_GROUP_BITS(group->channel_ios));
+
+		/* set sampling I/O */
+		sys_set_bits((mem_addr_t)&config->tsc->IOSCR, GET_GROUP_BITS(group->sampling_io));
+
+		/* enable group */
+		if (!group->use_as_shield) {
+			sys_set_bit((mem_addr_t)&config->tsc->IOGCSR, group->group - 1);
+		}
+	}
+
+	/* disable interrupts */
+	sys_clear_bits((mem_addr_t)&config->tsc->IER, TSC_IER_EOAIE | TSC_IER_MCEIE);
+
+	/* clear interrupts */
+	sys_set_bits((mem_addr_t)&config->tsc->ICR, TSC_ICR_EOAIC | TSC_ICR_MCEIC);
+
+	/* enable peripheral */
+	sys_set_bit((mem_addr_t)&config->tsc->CR, TSC_CR_TSCE_Pos);
+
+	if (IS_ENABLED(CONFIG_STM32_TSC_INTERRUPT)) {
+		config->irq_func();
+	}
+
+	return 0;
+}
+
+#define STM32_TSC_GROUP_AND_COMMA(node_id)                                                         \
+	{                                                                                          \
+		.group = DT_PROP(node_id, group),                                                  \
+		.channel_ios = DT_PROP(node_id, channel_ios),                                      \
+		.sampling_io = DT_PROP(node_id, sampling_io),                                      \
+		.use_as_shield = DT_PROP(node_id, use_as_shield),                                  \
+	},
+
+#define STM32_TSC_INIT(index)                                                                      \
+	static const struct stm32_pclken pclken_##index[] = STM32_DT_INST_CLOCKS(index);           \
+                                                                                                   \
+	PINCTRL_DT_INST_DEFINE(index);                                                             \
+                                                                                                   \
+	static void stm32_tsc_irq_init_##index(void)                                               \
+	{                                                                                          \
+		IRQ_CONNECT(DT_INST_IRQN(index), DT_INST_IRQ(index, priority), stm32_tsc_isr,      \
+			    DEVICE_DT_INST_GET(index), 0);                                         \
+		irq_enable(DT_INST_IRQN(index));                                                   \
+	};                                                                                         \
+                                                                                                   \
+	static const struct stm32_tsc_group_config groups_cfg_##index[] = {                        \
+		DT_INST_FOREACH_CHILD_STATUS_OKAY(index, STM32_TSC_GROUP_AND_COMMA)};              \
```

Comment:
you could actually use `DT_INST_FOREACH_CHILD_STATUS_OKAY_SEP` and drop the comma from `STM32_TSC_GROUP_AND_COMMA` and then rename it to something else, then you could potentially reuse the define macro for both "foreach" and single instances, which you don't really need here but it's a nice pattern to follow